### PR TITLE
ViT architecture updates

### PR DESCRIPTION
**Pull request by pomonam**
This PR attempts to address #184 and #189.

To match the configuration from target-setting runs:
1. Change architecture configuration to `S/16` from `B/32`.
2. Use identical parameter initialization methods for both Jax and PyTorch workloads.
3. Clean up the code by removing any intermediate unused variables.
4. Add gradient clipping for PyTorch and use gradient clipping of 1.0 in the target-setting run.
5. Fix gradient clipping for Jax (I suspect there is a missing `sqrt` if I am not mistaken).

However, there are still some differences:
1. The way we are applying weight decay (e.g., apply to tensors with above or equal to 2 dimensions vs. apply to all tensors) -- which can change the optimal LR and WD.
2. Data augmentation (e.g., using RandAugment) -- which can change the optimal LR due to the implicit regularization effect.

### CODE REVIEW

File: algorithmic_efficiency/workloads/imagenet_vit/imagenet_jax/models.py [hunk 1]
```diff
@@ -95,68 +84,49 @@ class Encoder(nn.Module):
...
   """ViT model."""
 
-  num_classes: int
+  num_classes: int = 1000
   patch_size: Sequence[int] = (16, 16)
   width: int = 768
   depth: int = 12
   mlp_dim: Optional[int] = None  # Defaults to 4x input dim
   num_heads: int = 12
-  posemb: str = 'sincos2d'  # Can also be "learn"
```
Comment (runame) [inline algorithmic_efficiency/workloads/imagenet_vit/imagenet_jax/models.py:147]: General question: should we remove the code for all options which we don't use, to simplify the code?

File: algorithmic_efficiency/workloads/imagenet_vit/imagenet_jax/models.py [hunk 2]
```diff
@@ -95,68 +84,49 @@ class Encoder(nn.Module):
...
+  posemb: str = 'sincos2d'
   rep_size: Union[int, bool] = True
   dropout_rate: Optional[float] = 0.0  # If None, defaults to 0.0.
-  pool_type: str = 'gap'  # Can also be 'map' or 'tok'
+  pool_type: str = 'gap'  # Can also be 'tok'
```
Comment (runame) [inline algorithmic_efficiency/workloads/imagenet_vit/imagenet_jax/models.py:111]: Same question as above.

File: algorithmic_efficiency/workloads/imagenet_vit/imagenet_pytorch/models.py [hunk 3]
```diff
@@ -218,9 +199,9 @@ def __init__(
       depth: int = 12,
       mlp_dim: Optional[int] = None,  # Defaults to 4x input dim
       num_heads: int = 12,
-      posemb: str = 'sincos2d',  # Can also be 'learn'
+      posemb: str = 'sincos2d',
```
Comment (runame) [inline algorithmic_efficiency/workloads/imagenet_vit/imagenet_pytorch/models.py:202]: Same question as above.

File: algorithmic_efficiency/workloads/imagenet_vit/imagenet_pytorch/models.py [hunk 4]
```diff
@@ -218,9 +199,9 @@ def __init__(
...
       rep_size: Union[int, bool] = True,
-      dropout: float = 0.0,
+      dropout_rate: float = 0.0,
       pool_type: str = 'gap',  # Can also be 'tok'
```
Comment (runame) [inline algorithmic_efficiency/workloads/imagenet_vit/imagenet_pytorch/models.py:205]: Same question as above.

**Review by runame — APPROVED** [review]
LGTM

File: algorithmic_efficiency/workloads/imagenet_vit/imagenet_jax/models.py (hunk 1 above)
Comment (pomonam) [reply algorithmic_efficiency/workloads/imagenet_vit/imagenet_jax/models.py:147]: Yes, we can remove the code for all options for simplicity. I will work on it.

Comment (runame) [reply algorithmic_efficiency/workloads/imagenet_vit/imagenet_jax/models.py:147]: Thanks!

**Comment by znado** [comment]
Regarding your question about which tensors to apply WD to, I think it's OK to apply it to all because [here](https://github.com/google/init2winit/blob/master/init2winit/trainer_lib/trainer.py#L110) in init2winit we apply it to all (the rank threshold is 0 which is all tensors), and it looks like PyTorch also applies it to all tensors. The config for ViT is confusing because in init2winit we support both L2 regularization and WD, and so the `"l2_decay_rank_threshold": 2,` is ignored because we have `"l2_decay_factor": null,`, and then WD is applied because we have `"weight_decay": 0.026595371563300656`.

Regarding randaug, that is a real diff that we should implement, I commented on https://github.com/mlcommons/algorithmic-efficiency/issues/184. Can you make a follow-up PR that adds it? I'm not sure the best place to grab a PyTorch implementation, but you can use the tf.data one from init2winit [here](https://github.com/google/init2winit/blob/master/init2winit/dataset_lib/imagenet_dataset.py#L264).

**Comment by znado** [comment]
Thanks so much for hunting down all these diffs. These were major changes that we needed to resolve, and finding them requires talented ML and engineering skills while carefully reading the code.
